### PR TITLE
JSON.parse(), JSON.stringify() more specific declarations for #6955

### LINKED ISSUE #6955: More specific JSON parse reviver argument in lib.d.ts
Currently, the JSON interface in lib.d.ts declares the following:

```
parse(text: string, reviver?: (key: any, value: any) => any): any;
```

However, my understanding is that the spec guaratees a string for the key (see http://www.ecma-international.org/ecma-262/5.1/#sec-15.12.2 paragraph "The abstract operation Walk ... String name ..."), in consequence it could be more specific:

```
parse(text: string, reviver?: (key: string, value: any) => any): any;
```

### PR DESCRIPTION
**Pull request by AnyhowStep**
Fixes #6955 

I made a few comments in that issue about being unclear about how the whole PR process works,

And it seems I broke 5 tests,

image
image
image

I assume it's because they are expecting the previous signature for `JSON.parse()` and `JSON.stringify()` but I've changed them.

I'm not sure how to read the output files in `tests/baselines/local` and decide if they're okay to `baseline-accept` (which I assume updates the expected results) but I made a best-effort attempt to look through them and they seemed okay to me.

### CODE REVIEW

**Comment by AnyhowStep** [comment]
@mhegazy

I might be a little hasty and excited but could this be reviewed?

**Comment by dalen** [comment]
Shouldn't `JSON.parse` really return an `unknown` rather than `any`?

**Comment by AnyhowStep** [comment]
Ah, yes. That makes sense. Personally, I'm in favour of it now that you've pointed it out but I think it might break code for a lot of people.

File: lib/lib.es5.d.ts [hunk 1]
```diff
@@ -1046,14 +1046,14 @@ interface JSON {
...
-    parse(text: string, reviver?: (key: any, value: any) => any): any;
+    parse(text: string, reviver?: (this : any, key: string, value: any) => any): any;
     /**
       * Converts a JavaScript value to a JavaScript Object Notation (JSON) string.
       * @param value A JavaScript value, usually an object or array, to be converted.
       * @param replacer A function that transforms the results.
       * @param space Adds indentation, white space, and line break characters to the return-value JSON text to make it easier to read.
       */
-    stringify(value: any, replacer?: (key: string, value: any) => any, space?: string | number): string;
+    stringify(value: any, replacer?: (this : any, key: string, value: any) => any, space?: string | number): string;
```
Comment (DanielRosenwasser) [inline lib/lib.es5.d.ts]: Can you remove the stray space?

**Comment by RyanCavanaugh** [comment]
@typescript-bot please this this, I'm paranoid now

**Comment by RyanCavanaugh** [comment]
@typescript-bot please test this?

**Comment by RyanCavanaugh** [comment]
@typescript-bot test this please

**Comment by RyanCavanaugh** [comment]
Nothing relevant in RWC, GTG

**Pull request merged**